skipper_test.go

```diff
@@ -384,6 +385,266 @@ func Example_ratelimitRegistryBinding() {
+	})
+	defer reg.Close()
+
+	docFmt := `
+r1: * -> enableAccessLog(4,5) -> fifo(100,100,"3s") -> <roundRobin, "%s", "%s", "%s">;
+r2: PathRegexp("/endpoints") -> enableAccessLog(2,4,5) -> fifo(100,100,"3s") -> <roundRobin, "%s", "%s", "%s">;
+`
+	docApiserver := fmt.Sprintf(docFmt, u1.String(), u2.String(), u3.String(), u1.String(), u2.String(), u3.String())
+	dc, err := testdataclient.NewDoc(docApiserver)
+	if err != nil {
+		t.Fatalf("Failed to create testdataclient: %v", err)
+	}
+
+	// create LB in front of apiservers to be able to switch the data served by apiserver
+	ro := routing.Options{
+		SignalFirstLoad: true,
+		FilterRegistry:  fr,
+		DataClients:     []routing.DataClient{dc},
+		PostProcessors: []routing.PostProcessor{
+			loadbalancer.NewAlgorithmProvider(),
+			reg,
+		},
+		SuppressLogs: true,
+	}
+	rt := routing.New(ro)
+	defer rt.Close()
+	<-rt.FirstLoad()
+	tracer := &tracingtest.Tracer{}
+	pr := proxy.WithParams(proxy.Params{
+		Routing:     rt,
+		OpenTracing: &proxy.OpenTracingParams{Tracer: tracer},
+	})
+	defer pr.Close()
+	lb := stdlibhttptest.NewServer(pr)
+	defer lb.Close()
+
+	rsvo := routesrv.Options{
+		Address:                         ":8082",
+		KubernetesURL:                   lb.URL,
+		KubernetesRedisServiceNamespace: "skipper",
+		KubernetesRedisServiceName:      "redis",
+		KubernetesHealthcheck:           true,
+		SourcePollTimeout:               1500 * time.Millisecond,
+	}
+
+	go routesrv.Run(rsvo)
+
+	for {
+		rsp, _ := http.DefaultClient.Head("http://localhost:8082/routes")
+		if rsp != nil && rsp.StatusCode == 200 {
+			break
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+
+	// run skipper proxy that we want to test
+	o := Options{
+		Address:                        ":9090",
+		EnableRatelimiters:             true,
+		EnableSwarm:                    true,
+		Kubernetes:                     true,
+		SwarmRedisEndpointsRemoteURL:   "http://localhost:8082/swarm/redis/shards",
+		KubernetesURL:                  lb.URL,
+		KubernetesHealthcheck:          true,
+		SourcePollTimeout:              1500 * time.Millisecond,
+		WaitFirstRouteLoad:             true,
+		ClusterRatelimitMaxGroupShards: 2,
+		SwarmRedisDialTimeout:          100 * time.Millisecond,
+		SuppressRouteUpdateLogs:        false,
+		SupportListener:                ":9091",
+		testOptions: testOptions{
+			redisUpdateInterval: time.Second,
+		},
+	}
+
+	sigs := make(chan os.Signal, 1)
+	go run(o, sigs, nil)
+
+	// wait for proxy being ready
+	for {
+		rsp, _ := http.DefaultClient.Get("http://localhost:9090/kube-system/healthz")
+		if rsp != nil && rsp.StatusCode == 200 {
+			break
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+
+	rate := 10
+	sec := 5
+	va := httptest.NewVegetaAttacker("http://localhost:9090/test", rate, time.Second, time.Second)
+	va.Attack(io.Discard, time.Duration(sec)*time.Second, "mytest")
+	t.Logf("Success [0..1]: %0.2f", va.Success())
+
+	if successRate := va.Success(); successRate < 0.1 || successRate > 0.5 {
```

Comment:
here we check success rate is between 10% and 50%: interval notation: successRate in (0.1, 0.5).
total requests 10req*5s=50
filter allows 1RPS, so 5 allowed requests in 5s
Expected successRate: 5/50=1/10=0.1
So why do we test (0.1, 0.5) as success?
I would rather say we should test for 0.1 +/- \epsilon where you can tune \epsilon such that the test is not flaky, but shows that the test makes sense. So maybe \epsilon=0.05 for example